### PR TITLE
Fix #11185: cache pretyped argument with adaptation in FunProto

### LINKED ISSUE #11185: Parameter untupling and automatic eta-expansion don't always mix
## Minimized code

```Scala
def foo(a: Int, b: Int) = a + b

Map(1 -> 2).map(foo _) // ok: List(3)

Map(1 -> 2).map(foo) // error
```

## Output

```scala
Map(1 -> 2).map(foo)
^^^^^^^^^^^^^^^
None of the overloaded alternatives of method map in trait IterableOps with types
 [B](f: ((Int, Int)) => B): scala.collection.immutable.Iterable[B]
 [K2, V2](f: ((Int, Int)) => (K2, V2)): Map[K2, V2]
match arguments ((foo : (a: Int, b: Int): Int))
```

## Expectation

Since the compiler can pick the right overload with explicit eta-expansion, I don't see why he can't do that with automatic eta-expansion.

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #11185: cache pretyped argument with adaptation in FunProto

### CODE REVIEW

File: tests/pos/i11185.scala [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Test:
+  def foo(a: Int, b: Int) = a + b
+
+  Map(1 -> 2).map(foo _)
```
Comment (Jasper-M) [inline tests/pos/i11185.scala:4]: Just a random thought, but if the collections ever get rewritten again `map` might not be overloaded anymore.

Comment (liufengyun) [reply tests/pos/i11185.scala:4]: Good point, I'll add another test with overloading.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 2]
```diff
@@ -2101,7 +2101,9 @@ trait Applications extends Compatibility {
             else defn.FunctionOf(commonParamTypes, WildcardType)
           overload.println(i"pretype arg $arg with expected type $commonFormal")
           if (commonParamTypes.forall(isFullyDefined(_, ForceDegree.flipBottom)))
-            withMode(Mode.ImplicitsEnabled)(pt.typedArg(arg, commonFormal))
+            withMode(Mode.ImplicitsEnabled) {
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala:2104]: At first glance this looks too simplistic. We do a complicated dance not to overcommit by caching arguments that can be passed to several different expected types. How can we be sure that we can simply cache in this case? Maybe we can, but we'd need a detailed comment why.

**Comment by odersky** [comment]
I just checked: it does not work with master anymore. I think https://github.com/lampepfl/dotty/pull/12138 is to blame.

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 2 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala:2104]: Yes, a comment is deserved here. Currently I see the rationale is this: the `expected type` for pretyping here come from all the overloading candidates and they are all fully defined.

> I just checked: it does not work with master anymore. I think #12138 is to blame.

I rebased against master, it seems to work.

**Review by odersky — APPROVED** [review]

**Pull request merged**